File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int cur = 0;
    int count = 0;
    public int kthSmallest(TreeNode root, int k) {
        dfs(root, k);

        return cur;
    }
    int dfs(TreeNode node, int k){
        if(node == null){
            return 0;
        }
        dfs(node.left, k);

        count++;

        if(count == k){
            cur = node.val;
        }

        return dfs(node.right, k);
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Codec {
public:
    int idx = 0;
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (root == nullptr) {
            return "#";
        }
        string val = to_string(root->val) + ",";
        return val + serialize(root->left) + "," + serialize(root->right);  
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if(data.length() == 0 or data[idx] == '#') {
            idx++;
            return nullptr;
        }

        string value;
        while(data[idx] != ',') {
            value += data[idx];
            idx++;
        }
        TreeNode* node = new TreeNode(stoi(value));
        idx++;
        node->left = deserialize(data);
        idx++;
        node->right = deserialize(data);

        return node;
    }
};
